dcrdtest/rpc_harness_test.go

```diff
@@ -658,3 +656,135 @@ func TestHarness(t *testing.T) {
 		}
 	}
 }
+
+// loggerWriter is an slog backend that writes to a test output.
+type loggerWriter struct {
+	l testing.TB
+}
+
+func (lw loggerWriter) Write(b []byte) (int, error) {
+	bt := bytes.TrimRight(b, "\r\n")
+	lw.l.Logf(string(bt))
+	return len(b), nil
+}
+
+// TestSetupTeardown tests that setting up and tearing down an rpc harness does
+// not leak any goroutines.
+func TestSetupTeardown(t *testing.T) {
+	// Add logging to ease debugging this test.
+	lw := loggerWriter{l: t}
+	bknd := slog.NewBackend(lw)
+	UseLogger(bknd.Logger("TEST"))
+	log.SetLevel(slog.LevelDebug)
+	defer UseLogger(slog.Disabled)
+
+	params := chaincfg.RegNetParams()
+	mainHarness, err := New(t, params, nil, nil)
+	if err != nil {
+		t.Fatalf("unable to create main harness: %v", err)
+	}
+
+	// Perform the setup.
+	ctx := testctx.WithTimeout(t, time.Second*30)
+	if err = mainHarness.SetUp(ctx, true, 2); err != nil {
+		t.Fatalf("unable to setup test chain: %v", err)
+	}
+
+	// Perform the teardown.
+	if err := mainHarness.TearDown(); err != nil {
+		t.Fatalf("unable to TearDown test chain: %v", err)
+	}
+
+	// There should be only 2 goroutines live.
+	prof := pprof.Lookup("goroutine")
+	gotCount, wantCount := prof.Count(), 2
```

Comment:
This is a technique I've not seen before.  I typically do this with channels and/or waitgroups.

I don't have any big objection to it, but this is probably fairly fragile since I suspect it relies on the semantics of the testing framework creating 2 goroutines.

Maybe it should get the count prior to calling `SetUp` and then ensuring it matches that value here so it's robust against framework changes?

Reply:
Point taken. Updating to use before/after count of goroutines. The "2" is indeed due to how the test runtime spawns the individual test functions.